File: controllers/sectionProps.py
# import matplotlib.pyplot as plt, mpld3
# import numpy as np
# from matplotlib.patches import Rectangle
# from matplotlib.collections import PatchCollection
import pandas as pd
import json
import sys
import ast


def initializeAnalysis(shapes):
    newshapes = []
    for x in shapes.split(" "):
        newshapes.append(ast.literal_eval(x))

    props = section_properties(newshapes)

    result = {
                "props": props,
                "geom": newshapes
            }

    print(json.dumps(result, indent=2))
    return result


def section_properties(test, mirror=False, xbar_init=0, ybar_init=0):
    columns = ["el", "base", "height", "Area", "xbar", "Ax", "Ax^2", "Ioy"]
    data_y = pd.DataFrame(columns=columns)

    # Y moment
    for i in range(len(test)):
        basey = test[i][2]
        heighty = test[i][3]
        areay = basey * heighty
        xbar = test[i][0] + basey / 2
        Ax = areay * xbar
        Ax_sq = Ax * xbar
        Ioy = (heighty * basey ** 3) / 12

        data_y.at[i, "el"] = i + 1
        data_y.at[i, "base"] = basey
        data_y.at[i, "height"] = heighty
        data_y.at[i, "Area"] = areay
        data_y.at[i, "xbar"] = xbar
        data_y.at[i, "Ax"] = Ax
        data_y.at[i, "Ax^2"] = Ax_sq
        data_y.at[i, "Ioy"] = Ioy

    # Sum all columns
    sums_y = data_y.sum(axis=0)
    xbar_cg = sums_y["Ax"] / sums_y["Area"]
    Ioy_all = sums_y["Ioy"] + sums_y["Ax^2"] - xbar_cg * sums_y["Ax"]

    columns = ["el", "base", "height", "Area", "ybar", "Ay", "Ay^2", "Iox"]
    data_x = pd.DataFrame(columns=columns)

    # X moment
    for i in range(len(test)):
        basex = test[i][2]
        heightx = test[i][3]
        areax = basex * heightx
        ybarx = test[i][1] + heightx / 2
        Ay_x = areax * ybarx
        Ay_sqx = Ay_x * ybarx
        Iox = (basex * heightx ** 3) / 12

        data_x.at[i, "el"] = i + 1
        data_x.at[i, "base"] = basex
        data_x.at[i, "height"] = heightx
        data_x.at[i, "Area"] = areax
        data_x.at[i, "ybar"] = ybarx
        data_x.at[i, "Ay"] = Ay_x
        data_x.at[i, "Ay^2"] = Ay_sqx
        data_x.at[i, "Iox"] = Iox

    # Sum all columns
    sums_x = data_x.sum(axis=0)
    ybar_cg = sums_x["Ay"] / sums_x["Area"]
    Iox_all = sums_x["Iox"] + sums_x["Ay^2"] - ybar_cg * sums_x["Ay"]

    sec_props = {
        "Iox": Iox_all,
        "Ioy": Ioy_all,
        "xbar": xbar_cg,
        "ybar": ybar_cg
    }

    return sec_props


# def main(args):
#     print("test")
#     print(args)
#     print(len(args))
#     print(args[1].split(" "))
# #     return initializeAnalysis([ast.literal_eval(args[1]), ast.literal_eval(args[1])])


if __name__ == "__main__":
#     main(sys.argv)
#     print("exit main")
#     print(sys.argv[1])
    initializeAnalysis(sys.argv[1])

# python sectionProps.py '[ 0.50, 0, 1.00, 0.30]' '[ 0.85, 0.3, 0.30, 2.40]' '[ 0, 2.7, 2.00, 0.30]'

# [(0.50,    0), 1.00, 0.30],
#         [(0.85,  0.3), 0.30, 2.40],
#         [(0   ,  2.7), 2.00, 0.30]

File: controllers/sectionProps.js
const ErrorResponse = require('../utils/errorResponse');
const asyncHandler = require('../middleware/async');
const SectionProp = require('../models/SectionProp');
const Part = require('../models/Parts');
const util = require('util');
const exec = util.promisify(require('child_process').exec);

// @desc      Get sectionProps
// @route     GET /api/v1/sectionProps
// @route     GET /api/v1/parts/:partId/sectionProps
// @access    Public
exports.getSectionProps = asyncHandler(async (req, res, next) => {
    if (req.params.partId) {
        const sectionProps = await SectionProp.find({part: req.params.partId})
        return res.status(200).json({
            success: true,
            count: sectionProps.length,
            data: sectionProps
        });
    } else {
        res.status(200).json(res.advancedResults);
    }
});

// @desc      Get sectionProp
// @route     GET /api/v1/sectionProps/:id
// @access    Public
exports.getSectionProp = asyncHandler(async (req, res, next) => {
    const sectionProp = await SectionProp.findById(req.params.id).populate({
        path: 'part',
        select: 'name description'
    });

    if(!sectionProp) {
        return next(new ErrorResponse(`No SectionProp with the id of ${req.params.id}`), 404);
    }

    res.status(200).json({
        success: true,
        data: sectionProp
    });
});

// @desc      Add sectionProp
// @route     POST /api/v1/parts/:partId/sectionProps
// @access    Private
exports.addSectionProp = asyncHandler(async (req, res, next) => {
    req.body.part = req.params.partId;
    req.body.user = req.user.id;

    const { sectionName} = req.body;
    const partId = req.params.partId;
    const part = await Part.findById(req.params.partId);

    if (!part) {
        return next(
            new ErrorResponse(
                `No part with the id of ${req.params.partId}`,
                404
            )
        );
    }

    // Make sure user is part owner
    if (part.user.toString() !== req.user.id && req.user.role !== 'admin') {
        return next(
            new ErrorResponse(
                `User ${req.user.id} is not authorized to add a section properties to part ${part._id}`,
                401
            )
        );
    }

    let result = await sectionTool(req.body.geom);

    // let result = await spawTool(req.body.geom)

    req.body = Object.assign(req.body, result.props)

    let sectionProp = await SectionProp.find({"sectionName":sectionName, "part":partId});

    if (sectionProp.length===0) {
        sectionProp = await SectionProp.create(req.body);
    } else {
        return next(
            new ErrorResponse(
                `Section Properties with name ${req.body.sectionName} already exist`,
                404
            )
        );
    }

    res.status(200).json({
        success: true,
        data: sectionProp
    });
});

// @desc      Update sectionProp
// @route     PUT /api/v1/sectionProps/:id
// @access    Private
exports.updateSectionProp = asyncHandler(async (req, res, next) => {
    let sectionProp = await SectionProp.findById(req.params.id);

    if (!sectionProp) {
        return next(
            new ErrorResponse(`No sectionProp with the id of ${req.params.id}`, 404)
        );
    }

    // Make sure user is section properties owner
    if (sectionProp.user.toString() !== req.user.id && req.user.role !== 'admin') {
        return next(
            new ErrorResponse(
                `User ${req.user.id} is not authorized to update section properties ${part._id}`,
                401
            )
        );
    }

    sectionProp = await SectionProp.findByIdAndUpdate(req.params.id, req.body, {
        new: true,
        runValidators: true
    });

    sectionProp.save();

    res.status(200).json({
        success: true,
        data: sectionProp
    });
});

// @desc      Delete sectionProp
// @route     DELETE /api/v1/sectionProps/:id
// @access    Private
exports.deleteSectionProp = asyncHandler(async (req, res, next) => {
    const sectionProp = await SectionProp.findById(req.params.id);

    if (!sectionProp) {
        return next(
            new ErrorResponse(`No sectionProp with the id of ${req.params.id}`, 404)
        );
    }

    // Make sure user is section properties owner
    if (sectionProp.user.toString() !== req.user.id && req.user.role !== 'admin') {
        return next(
            new ErrorResponse(
                `User ${req.user.id} is not authorized to delete section properties ${part._id}`,
                401
            )
        );
    }

    await sectionProp.remove();

    res.status(200).json({
        success: true,
        data: {}
    });
});

// async function spawTool(geom) {
//     const { spawn } = require('child_process');
//     const inputString = [];
//     for (i = 0; i < geom.length; i++) {
//         inputString.push("[" + geom + "]");
//     }
//
//     let resultDataSet = []
//     // const process = spawn('python', ['controllers/sectionProps.py', inputString.join(" ")]);
//
//     const { stdout, stderr } = await exec(`python controllers/sectionProps.py ${inputString.join(" ")}`);
//
//     if (stderr) {
//         console.error(`error: ${stderr}`);
//     }
//     // console.log(`Number of files ${stdout}`);
//
//     return JSON.parse(Buffer.from(stdout))
// };

async function sectionTool(geom) {

    let result = {
        "elemX": {
            "areay":0,
            "Ax":0,
            "Ax_sq":0,
            "Ioy":0},
        "elemY": {
            "areax":0,
            "Ay":0,
            "Ay_sq":0,
            "Iox":0},
        "props":{
            "Iox":0,
            "Ioy":0,
            "xbar":0,
            "ybar":0
        }
    }

    // Y moment
    for (i = 0; i < geom.length; i++) {
        // elemX.basey      += geom[i][2];
        // elemX.heighty    += geom[i][3];
        result.elemX.areay += geom[i][2] * geom[i][3];
        // elemX.xbar       += geom[i][0] + geom[i][2] / 2;
        result.elemX.Ax += (geom[i][2] * geom[i][3]) * (geom[i][0] + geom[i][2] / 2);
        result.elemX.Ax_sq += ((geom[i][2] * geom[i][3]) * (geom[i][0] + geom[i][2] / 2)) * (geom[i][0] + geom[i][2] / 2);
        result.elemX.Ioy += (geom[i][3] * geom[i][2] ** 3) / 12;
    }

    let xbar_cg = result.elemX.Ax / result.elemX.areay;
    let Ioy_all = result.elemX.Ioy + result.elemX.Ax_sq - xbar_cg * result.elemX.Ax;

    // X moment
    for (i = 0; i < geom.length; i++) {
        // elemY.basex      += geom[i][2];
        // elemY.heightx    += geom[i][3];
        result.elemY.areax += geom[i][2] * geom[i][3];
        // elemY.ybar       += geom[i][1] + geom[i][3] / 2;
        result.elemY.Ay += (geom[i][2] * geom[i][3]) * (geom[i][1] + geom[i][3] / 2);
        result.elemY.Ay_sq += ((geom[i][2] * geom[i][3]) * (geom[i][1] + geom[i][3] / 2)) * (geom[i][1] + geom[i][3] / 2);
        result.elemY.Iox += (geom[i][2] * geom[i][3] ** 3) / 12;
    }

    let ybar_cg = result.elemY.Ay / result.elemY.areax;
    let Iox_all = result.elemY.Iox + result.elemY.Ay_sq - ybar_cg * result.elemY.Ay;

    result.props.Iox  = Iox_all.valueOf();
    result.props.Ioy  = Ioy_all.valueOf();
    result.props.xbar = xbar_cg.valueOf();
    result.props.ybar = ybar_cg.valueOf();

    return result;
}

